### PR TITLE
feat: Outcome Sampling MCCFR, memoized enumeration, experiment framework

### PR DESCRIPTION
**Pull request by BedirT**
## Summary

- Outcome Sampling MCCFR (corrected per OpenSpiel, 2,300x faster than External on 3x3)
- Memoized exhaustive enumeration (7,000x speedup — 4x3 now feasible)
- Restructured `src/` into `game/` + `solver/` modules
- Experiment framework with persistent results (`experiments/` + `results/`)
- 76 tests (30 Rust + 46 Python)

## Reasoning Chain

**External Sampling** worked on 2x2/3x2 but hit two walls:
1. 3x3 too slow (7 iters/s) — O(|A|^d) branching
2. Missing info states (39/42) — zero-prob opponent actions never sampled

→ **Outcome Sampling**: O(depth) per iteration + epsilon-greedy exploration

**PR review** (Copilot + Codex) caught 3 importance weight bugs:
1. Return value missing `sigma/q` correction
2. Epsilon applied at opponent nodes (should be update player only)
3. Terminal returning utility/pi_sample (should return raw utility)

→ Aligned with **OpenSpiel formulation** — epsilon only at update player

Then: how do we *know* we found all states? Don't trust thesis numbers.
→ **Exhaustive enumeration** (`enumerate_game_tree`) for ground truth

But naive DFS took 6.2 hours on 3x3 (9.47B terminals).
→ **Memoization** on full game state: 3x3 in 3s, 4x3 in 8.5min

Then: where are results saved for the paper?
→ **Experiment framework**: scripts + markdown writeups + CSV/JSON data

## Ground Truth (All Independently Verified)

| Board | Info States | Game States | Time | Status |
|-------|-----------|-------------|------|--------|
| 2x2 | **42** | 105 | 0.001s | Thesis MATCH |
| 2x3 | **314** | 1,797 | 0.02s | NEW |
| 3x2 | **410** | 2,469 | 0.02s | Thesis MATCH |
| 3x3 | **12,556** | 283,859 | 3.0s | Thesis MATCH |
| 4x3 | **367,919** | 31,949,417 | 8.5min | Thesis MATCH |
| 3x4 | **341,033** | 28,560,489 | 7.1min | NEW |

## Test Plan
- [x] 30 Rust tests (game + solver + enumeration incl. 3x3)
- [x] 46 Python tests (engine + mccfr + enumeration)
- [x] `make check` clean
- [x] All 4 thesis values independently confirmed
- [x] 18 PR review comments addressed across 3 rounds

### CODE REVIEW

**Comment by BedirT** [comment]
@codex

**Comment by BedirT** [comment]
@claude review

File: src/solver/mccfr.rs [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+                }
+            }
+
+            // Average strategy: reach-weighted
+            let strat_weight = pi_i / pi_sample;
+            for i in 0..num_actions {
+                data.strategy_sum[i] += strat_weight * sigma[i];
+            }
+
+            tail * sigma[a_idx]
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs]: Fixed in d2227ef. Aligned with OpenSpiel: return `value_estimate = sigma[a] / q[a] * child_value`. Terminal now returns raw utility, weighting deferred to regret update with `opp_reach / sample_reach`.

File: src/solver/mccfr.rs [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+        Self {
+            info_states: HashMap::new(),
+            iterations: 0,
+            rows,
+            cols,
+            sampling: sampling.unwrap_or(Sampling::Outcome),
+            epsilon: epsilon.unwrap_or(0.6),
+            rng: SmallRng::seed_from_u64(seed.unwrap_or(42)),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs:118]: Fixed in d2227ef. Constructor validates `0 < epsilon <= 1`, returns PyValueError otherwise. Rust test added.

File: experiments/exp001_mccfr_verification.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        )
+        writer.writeheader()
+        for r in all_records:
+            row = {k: v for k, v in r.items() if k in writer.fieldnames}
+            writer.writerow(row)
+
+    # Save full JSON (includes strategies)
+    json_path = RESULTS_DIR / "full_results.json"
+    with open(json_path, "w") as f:
+        json.dump(all_records, f, indent=2, default=str)
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp001_mccfr_verification.py:110]: Fixed in d2227ef. Added `RESULTS_DIR.mkdir(parents=True, exist_ok=True)`.

File: experiments/exp002_info_state_enumeration.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    with open(csv_path, "w", newline="") as f:
+        writer = csv.DictWriter(f, fieldnames=list(records[0].keys()))
+        writer.writeheader()
+        writer.writerows(records)
+
+    # Save JSON
+    json_path = RESULTS_DIR / "full_results.json"
+    with open(json_path, "w") as f:
+        json.dump(records, f, indent=2)
+
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp002_info_state_enumeration.py:94]: Fixed in d2227ef. Same fix applied.

File: src/game/enumerate.rs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+
+    let player = state.rs_current_player();
+    let pi = player.index();
+    let info_key = state.rs_info_state_string(player);
+    info_states[pi].insert(info_key);
+
+    state.rs_legal_actions(actions_buf);
+    let actions: Vec<usize> = actions_buf.clone();
+
+    for &action in &actions {
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/game/enumerate.rs:130]: Acknowledged — apply/undo would speed up the 3x3 case significantly. Deferring to a future optimization pass. The 6.2h exhaustive run already completed and confirmed 12,556. Tracked as a known improvement.

File: src/solver/mccfr.rs [hunk 6]
```diff
@@ -0,0 +1,362 @@
...
+        let info_key = state.rs_info_state_string(player);
+        self.get_strategy(&info_key, num_actions, sigma_buf);
+        let sigma: Vec<f32> = sigma_buf.clone();
+        let actions: Vec<usize> = actions_buf.clone();
+
+        // Epsilon-on-policy sampling probabilities
+        let eps = self.epsilon;
+        let n = num_actions as f32;
+        q_buf.clear();
+        q_buf.extend(sigma.iter().map(|&s| eps / n + (1.0 - eps) * s));
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs]: Fixed in d2227ef. Constructor rejects epsilon outside (0, 1]. Rust test `epsilon_validation()` covers edge cases.

File: src/solver/mccfr.rs [hunk 7]
```diff
@@ -0,0 +1,362 @@
...
+                data.strategy_sum[i] += strat_weight * sigma[i];
+            }
+
+            tail * sigma[a_idx]
+        } else {
+            self.outcome_sampling(
+                &mut child,
+                update_player,
+                pi_i,
+                pi_opp * sigma[a_idx],
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs]: @codex Fixed in https://github.com/BedirT/darkhex/commit/d2227ef1c0294cbd71767e73069ee0205e2e7a14. Root cause: epsilon was applied at ALL nodes. Per OpenSpiel, epsilon-greedy applies ONLY at update player nodes. Opponent samples from sigma (q==sigma), no correction needed. Also fixed return value and terminal formulation.

**Comment by BedirT** [comment]
@codex review

File: experiments/exp002_info_state_enumeration.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        record = {
+            "board": board_key,
+            "rows": rows,
+            "cols": cols,
+            "total_info_states": stats.total_info_states,
+            "p0_info_states": stats.info_states_by_player[0],
+            "p1_info_states": stats.info_states_by_player[1],
+            "terminal_states": stats.terminal_states,
+            "max_depth": stats.max_depth,
+            "wall_time_s": round(wall_time, 3),
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp002_info_state_enumeration.py:61]: Fixed — regenerated results from the actual experiment scripts so schemas match. Both CSV and JSON now consistent with script output.

File: experiments/exp001_mccfr_verification.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            "info_states_found": solver.num_info_states(),
+            "info_states_ground_truth": ground_truth.total_info_states,
+            "coverage_pct": solver.num_info_states()
+            / ground_truth.total_info_states
+            * 100,
+            "wall_time_s": round(wall_time, 3),
+            "iters_per_sec": round(n_iters / wall_time, 0),
+            "p0_initial_strategy": p0_strat,
+            "p1_initial_strategy": p1_strat,
+        })
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp001_mccfr_verification.py:55]: Fixed — regenerated results from the experiment script. Schema now consistent between script and committed data.

File: src/solver/mccfr.rs [hunk 10]
```diff
@@ -0,0 +1,389 @@
...
+fn sample_action(probs: &[f32], rng: &mut SmallRng) -> usize {
+    let r: f32 = rng.gen();
+    let mut cum = 0.0;
+    for (i, &p) in probs.iter().enumerate() {
+        cum += p;
+        if r < cum {
+            return i;
+        }
+    }
+    probs.len() - 1
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs:68]: Fixed — sample_action now samples r in [0, sum(probs)) and fallback returns last nonzero-probability index. Prevents division-by-zero from floating-point rounding.

File: src/solver/mccfr.rs [hunk 11]
```diff
@@ -0,0 +1,389 @@
...
+            return Err(pyo3::exceptions::PyValueError::new_err(
+                "epsilon must be in (0, 1]",
+            ));
+        }
+        Ok(Self {
+            info_states: HashMap::new(),
+            iterations: 0,
+            rows,
+            cols,
+            sampling: sampling.unwrap_or(Sampling::Outcome),
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/solver/mccfr.rs]: Fixed — epsilon validation now only applies when sampling == Outcome. External Sampling accepts any epsilon value (ignored). Test added: `epsilon_ignored_for_external()`.

File: experiments/exp002_info_state_enumeration.md [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+- Compare computed counts against thesis values (Tapkan 2022)
+- Record per-player split, terminal count, max tree depth
+- Measure wall-clock time to assess feasibility for larger boards
+
+## Results
+
+| Board | Info States | P0 | P1 | Terminals | Max Depth | Time | Thesis | Match |
+|-------|-----------|-----|-----|-----------|-----------|------|--------|-------|
+| 2x2 | **42** | 17 | 25 | 216 | 7 | 0.001s | 42 | MATCH |
+| 2x3 | **314** | 134 | 180 | 53,920 | 11 | 0.12s | N/A | NEW |
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp002_info_state_enumeration.md]: Fixed — corrected to P0=147, P1=167 (verified by `enumerate_game_tree(2, 3)`).

File: tests/test_enumerate.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+        solver.solve(10000)
+        assert solver.num_info_states() == expected
+
+    def test_outcome_3x2_high_coverage(self):
+        """3x2: expect >95% coverage (some opponent zero-prob paths missed)."""
+        expected = enumerate_game_tree(3, 2).total_info_states
+        solver = MCCFRSolver(3, 2, Sampling.Outcome, epsilon=0.6, seed=42)
+        solver.solve(100000)
+        coverage = solver.num_info_states() / expected
+        assert coverage > 0.95, f"coverage {coverage:.1%} below 95%"
```
[the remark this answers is not in the document]
Comment (BedirT) [reply tests/test_enumerate.py]: Fixed in b491ce3. Reduced to 20k iterations / 85% threshold.

File: src/lib.rs [hunk 14]
```diff
@@ -17,6 +16,9 @@ fn _engine(m: &Bound<'_, PyModule>) -> PyResult<()> {
     m.add_class::<CollisionRule>()?;
     m.add_class::<CollisionInfo>()?;
     m.add_class::<DarkHexState>()?;
+    m.add_class::<Sampling>()?;
     m.add_class::<MCCFRSolver>()?;
+    m.add_class::<GameTreeStats>()?;
+    m.add_function(wrap_pyfunction!(enumerate_game_tree, m)?)?;
```
[the remark this answers is not in the document]
Comment (BedirT) [reply src/lib.rs:22]: It compiles — `wrap_pyfunction!` is re-exported by `pyo3::prelude::*` which is already imported. Verified: 29 Rust tests pass.

File: experiments/exp001_mccfr_verification.md [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+## Results
+
+### Info State Coverage
+
+| Board | Ground Truth | 1k iters | 10k | 100k | 500k |
+|-------|-------------|----------|-----|------|------|
+| 2x2 | 42 | 42 (100%) | 42 | 42 | — |
+| 3x2 | 410 | 398 (97.1%) | 410 (100%) | 410 | — |
+| 3x3 | 12,556 | 4,479 (35.7%) | 10,368 (82.6%) | 12,507 (99.6%) | **12,556 (100%)** |
+
```
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp001_mccfr_verification.md:28]: Fixed in b491ce3. Table now matches the committed convergence.csv from the corrected OS-MCCFR (epsilon at update player only).

File: docs/algorithms/mccfr.md [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+- Zinkevich, M. et al. (2007). "Regret Minimization in Games with Incomplete Information." NeurIPS.
+- Lanctot, M. et al. (2009). "Monte Carlo Sampling for Regret Minimization in Extensive Games." NeurIPS.
+- Neller, T. & Lanctot, M. (2013). "An Introduction to Counterfactual Regret Minimization."
+
+## Computational Complexity
+
+| Variant | Time/iteration | Space | Info state coverage |
+|---------|---------------|-------|---------------------|
+| External | O(\|A\|^d) | O(\|I\|) | Misses zero-prob branches |
+| Outcome | O(d) | O(\|I\|) | Full (epsilon-greedy) |
```
[the remark this answers is not in the document]
Comment (BedirT) [reply docs/algorithms/mccfr.md]: Fixed in b491ce3. Changed to 'High (epsilon at update player only; asymptotically full)'. The corrected formulation per OpenSpiel applies epsilon exploration only at update player nodes, so opponent zero-prob branches are not visited.

File: docs/algorithms/mccfr.md [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+| Board | Exact Count | Method | Time |
+|-------|-------------|--------|------|
+| 2x2 | **42** (P0=17, P1=25) | Exhaustive DFS | 0.001s |
+| 3x2 | **410** (P0=172, P1=238) | Exhaustive DFS | 0.23s |
+| 3x3 | **12,556** (P0=6293, P1=6263) | Exhaustive DFS | 6.2h |
+
+All three verified by exhaustive enumeration. 3x3 has 9.47 billion terminal
+histories (max depth 17 due to CDH collision retries), explaining the 6.2h
+runtime. MCCFR Outcome Sampling reaches the same 12,556 in just 26 seconds.
+
```
[the remark this answers is not in the document]
Comment (BedirT) [reply docs/algorithms/mccfr.md:94]: Fixed in b491ce3. Removed the stale claim. Corrected OS-MCCFR reaches 11,308/12,556 (90.1%) at 500k iterations. The old numbers were from the buggy implementation (epsilon at all nodes).

File: tests/test_enumerate.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (BedirT) [reply tests/test_enumerate.py]: Fixed in b491ce3. Reduced to 20k iters / 85% threshold.

File: src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (BedirT) [reply src/lib.rs:22]: Compiles fine — wrap_pyfunction! is re-exported by pyo3::prelude::*. 29 Rust tests pass.

File: experiments/exp001_mccfr_verification.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (BedirT) [reply experiments/exp001_mccfr_verification.md:28]: Fixed in b491ce3. Table now matches committed convergence.csv.

File: docs/algorithms/mccfr.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (BedirT) [reply docs/algorithms/mccfr.md]: Fixed in b491ce3. Changed to 'High (epsilon at update player only; asymptotically full)'.

File: docs/algorithms/mccfr.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (BedirT) [reply docs/algorithms/mccfr.md:94]: Fixed in b491ce3. Removed stale claim. Corrected OS-MCCFR reaches 11,308/12,556 (90.1%) at 500k.

File: tests/test_mccfr.py [hunk 18]
```diff
@@ -1,65 +1,84 @@
...
+        solver.solve(100)
+        assert solver.iterations() == 100
+        assert solver.num_info_states() > 0
+
+    def test_2x2_discovers_all_info_states(self):
+        """Epsilon-greedy exploration should find all 42 info states."""
+        solver = MCCFRSolver(2, 2, Sampling.Outcome, epsilon=0.6, seed=42)
+        solver.solve(10000)
+        n = solver.num_info_states()
+        assert n >= 40, f"expected >=40, got {n}"
```
[the remark this answers is not in the document]
Comment (BedirT) [reply tests/test_mccfr.py:55]: Good catch. The test does find all 42 deterministically with seed=42, but the assertion is loose as a safety margin. Updated docstring to match the relaxed assertion.
